common.sh

```diff
@@ -188,7 +196,18 @@ if [[ "${IP_STACK}" = "v6" ]]; then
 else
   export PROVISIONING_HOST_EXTERNAL_IP=${PROVISIONING_HOST_EXTERNAL_IP:-$(python -c "import ipaddress; print(next(ipaddress.ip_network(u\"$EXTERNAL_SUBNET_V4\").hosts()))")}
 fi
-export MIRROR_IP=${MIRROR_IP:-$PROVISIONING_HOST_IP}
+export MIRROR_IP=${MIRROR_IP:-$PROVISIONING_HOST_EXTERNAL_IP}
+
+# When the provisioning network is disabled, we use IP's on the external
+# network for the provisioning IP's:
+if [[ "$PROVISIONING_NETWORK_PROFILE" == "Disabled" ]]; then
+  # TODO(stbenjam): Calculate these, support IPv4
+  export BOOTSTRAP_PROVISIONING_IP="fd2e:6f44:5dd8:c956::7"
+  export CLUSTER_PROVISIONING_IP="fd2e:6f44:5dd8:c956::8"
```

Comment:
So this effectively adds a requirement for 2 more VIP-like reservations outside of the DHCP pool.

Perhaps that's unavoidable but I'm trying to think if there's any way around it, hmm.

Reply:
What if we used a similar approach to CRIO and kubelet e.g:

https://github.com/openshift/machine-config-operator/blob/master/templates/common/baremetal/units/nodeip-configuration.service.yaml

https://github.com/openshift/baremetal-runtimecfg/blob/master/scripts/nodeip-finder

This creates files like:

$ cat /etc/systemd/system/kubelet.service.d/20-nodenet.conf
[Service]
Environment="KUBELET_NODE_IP=fd2e:6f44:5dd8:c956::14"

We could write a similar file for the ironic.service, and use that on both the bootstrap and cluster to determine the controlplane IP for the ironic config?

@celebdor any thoughts on that idea?

Reply:
On the cluster I guess it would be an initContainer with host networking, and on the bootstrap we could run it directly like (inside startironic.sh):

https://github.com/openshift/machine-config-operator/blob/master/templates/common/baremetal/files/NetworkManager-resolv-prepender.yaml#L31

Reply:
I explicitly called this as out of scope on the enhancement request, I feel like that discussion should've happened then. 

Keep in mind, the reason I have it like this is both provisioning infrastructures are online at the same time.  The bootstrap ironic can still be provisioning the third control plane node, while the cluster has already brought up machine-api-operator and is provisioning workers.

Reply:
https://github.com/openshift/enhancements/blob/dda6f2b7a284876ff666b1e77b8ed1ee65fbd612/enhancements/baremetal/baremetal-provisioning-optional.md#non-goals

https://github.com/openshift/enhancements/blob/dda6f2b7a284876ff666b1e77b8ed1ee65fbd612/enhancements/baremetal/baremetal-provisioning-optional.md#implementation-detailsnotesconstraints

Reply:
> We could write a similar file for the ironic.service, and use that on both the bootstrap and cluster to determine the controlplane IP for the ironic config?

In this case you would be using the DHCP addresses on the control plane network? How would you get this data back to the installer? The IP of Ironic has to be knowable before terraform even starts.

Reply:
> @hardys The nodeip-configuration.service helps you find the IPs, but you'd still need to reserve fd2e:6f44:5dd8:c956::7 and fd2e:6f44:5dd8:c956::8, wouldn't you?

No I'm suggesting we just use the controlplane IP instead, which may be reservations (inside the pool) for the cluster nodes, but probably not for the bootstrap VM.  The main point is we avoid extra steps for the human user (another 2 ips to reserve that must be outside of the DHCP pool).

@stbenjam I'd missed that was a non-goal, but given that e.g the assisted team are going to great lengths to avoid needing any VIP reservations at all I wanted to point it out.  It's something we can defer and revisit, but I feel like we have some of the pieces needed already in place so it'd be good to think about how it could potentially work.

Reply:
> > We could write a similar file for the ironic.service, and use that on both the bootstrap and cluster to determine the controlplane IP for the ironic config?
> 
> In this case you would be using the DHCP addresses on the control plane network? How would you get this data back to the installer? The IP of Ironic has to be knowable before terraform even starts.

Yeah I'd forgotten the current installer implementation makes it difficult to discover the bootstrap VM IP dynamically so it's probably more work than I initially thought.

For the cluster I guess it's simpler since the BMO can reach ironic via localhost in the same pod, so we'd just need to influence the ironic confg having discovered the IP to use?

Reply:
> @stbenjam I'd missed that was a non-goal, but given that e.g the assisted team are going to great lengths to avoid needing any VIP reservations at all I wanted to point it out.

I tried hard to communicate out this design decision in the enhancement request that was up for some time to review, as well as mentioning it more than once on Slack to the assisted installer folks. It's unfortunate this is coming up now. :-( 

> For the cluster I guess it's simpler since the BMO can reach ironic via localhost in the same pod, so we'd just need to influence the ironic confg having discovered the IP to use?

Removing the cluster IP is probably more doable than trying to do it for the bootstrap. The installer creates all the hosts including workers as part of its asset generation, which fills in all the details like URL's for images, right? We'll need to trigger updates of the BMH's to correct that data once the Metal3 pod is up somewhere and you can do the node IP detection. You'll need to do that any time the Metal3 pod moves, as well. What would do that, CAPBM? 

So, what would this look like? Can MAO get the IP of the node the Metal3 pod is running on? Can MAO then pass it into CAPBM when it starts, and have it update all the BMH's?

Reply:
Could we take advantage of the fact that we're using the external network and have the MAO define a Service instead of probing for IP addresses? That would let us avoid reconfiguring the CAPBM pod every time metal3 is scheduled to a new host. Do Services work with pods running on the host network?

Reply:
> So, what would this look like? Can MAO get the IP of the node the Metal3 pod is running on? Can MAO then pass it into CAPBM when it starts, and have it update all the BMH's?

I think we can use baremetal-runtimecfg to get the IP which can then be used as input to the ironic containers.

> Could we take advantage of the fact that we're using the external network and have the MAO define a Service instead of probing for IP addresses?

That sounds nice, but I'm not sure if e.g IPA can resolve the service name, particularly in the case where we've not got dnsmasq providing a DNS via DHCP?

@stbenjam I certainly didn't mean to derail discussion of the current iteration by raising this now, given the scope of the enhancement it sounds fine to defer this for now and continue figuring out how it would work.

Reply:
If we use a configuration value that isn't static (name or IP) then every time the metal3 pod is rescheduled we have to reconfigure all of the MachineSets with the URL that includes the new value. We typically only have one, but it's possible to create more. I don't know that we have anything watching for the reschedule event today.

We don't want to change existing hosts or machines, because that will trigger reprovisioning.

Reply:
> If we use a configuration value that isn't static (name or IP) then every time the metal3 pod is rescheduled we have to reconfigure all of the MachineSets with the URL that includes the new value.

If we can get the name to resolve in IPA then moving the image cache to a service sounds preferable, but if not we could use the ingress VIP IP I guess since that won't change, even if the service gets rescheduled?

Reply:
Perhaps we should shift this discussion to a new enhancement, which could be a potential follow-up to this work?

I'm happy to help help draft that if it avoids distracting efforts to get this iteration done?

Reply:
Does the ingress VIP move with the metal3 pod?

Moving to an enhancement is a good idea.

Reply:
How does MCS do what it does? You can hit that by IP.

For services, don't you need to use the .apps address to make that work? This is getting into territory on OpenShift I don't understand well, but if we have the option of making it work this way... it'd be more elegant for sure than making people reserve an IP.

Reply:
> Does the ingress VIP move with the metal3 pod?

No, it's backed by the OpenShift routers for application traffic, so unrelated to the metal3 pod - my idea would depend on converting at least the webserver part that hosts the RHCOS image to a Service so we can reach it via the ingress VIP, but I'm not sure if that means we can only reach it by name, I guess probably it does but I'm not super familiar with how the application routing works atm.

> Moving to an enhancement is a good idea.

I'll do some investigation and start one, then we can consider if it's something we'd like to tackle during 4.7, and sorry for the tangent here @stbenjam :)